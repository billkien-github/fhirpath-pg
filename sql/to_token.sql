drop table if exists as_token_test;
create table as_token_test (resource jsonb);

insert into as_token_test (resource) values (
$$
{
  "resourceType": "Patient",
  "id": "example",
  "identifier": [
   {
    "use": "usual",
    "type": { "coding": [
      {"system": "sys1", "code": "code1"},
      {"system": "sys2", "code": "code2"}
    ]},
    "system": "isys",
    "value": "value1",
    "period": {"start": "2001-05-06"},
    "assigner": {"display": "Acme Healthcare"}
   },
   {
   "use": "official",
   "type": { "coding": [
     {"system": "itsys", "code": "itcode"}
   ]},
   "system": "isys2",
   "value": "ival2"
   }
  ],
  "organization": {
     "reference": "Organization/1"
  },
  "quantity": {
     "code": "code",
     "system": "system",
     "unit": "unit"
  },
  "telecom": [
      {
        "system": "phone",
        "value": "(03) 5555 6473",
        "use": "work"
      },
      {
        "system": "email",
        "value": "ivan@mail.ru",
        "use": "home"
      }
  ]

}
$$
);

SELECT fhirpath_as_token(resource, '.identifier', 'Identifier') from as_token_test;
SELECT fhirpath_as_token(resource, '.telecom', 'ContactPoint') from as_token_test;
SELECT fhirpath_as_token(resource, '.telecom.value', 'string') from as_token_test;
SELECT fhirpath_as_token(resource, '.telecom.use', 'code') from as_token_test;
SELECT fhirpath_as_token(resource, '.telecom.use', 'uri') from as_token_test;
SELECT fhirpath_as_token(resource, '.telecom.use', 'string') from as_token_test;
SELECT fhirpath_as_token(resource, '.identifier.type.coding', 'Coding') from as_token_test;
SELECT fhirpath_as_token(resource, '.organization', 'Reference') from as_token_test;
SELECT fhirpath_as_token(resource, '.quantity', 'Quantity') from as_token_test;
SELECT fhirpath_as_token(resource, '.identifier.type', 'CodeableConcept') from as_token_test;

SELECT fhirpath_as_token('{"a": true}', '.a', 'boolean') from as_token_test;
SELECT fhirpath_as_token('{"a": false}', '.a', 'boolean') from as_token_test;
SELECT fhirpath_as_token('{"a": "ups"}', '.a', 'boolean') from as_token_test;
SELECT fhirpath_as_token('{"a": 1234}', '.a', 'integer') from as_token_test;
SELECT fhirpath_as_token('{"a": 1234.1}', '.a', 'integer') from as_token_test;
