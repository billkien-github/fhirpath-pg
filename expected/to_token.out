drop table if exists as_token_test;
NOTICE:  table "as_token_test" does not exist, skipping
create table as_token_test (resource jsonb);
insert into as_token_test (resource) values (
$$
{
  "resourceType": "Patient",
  "id": "example",
  "identifier": [
   {
    "use": "usual",
    "type": { "coding": [
      {"system": "sys1", "code": "code1"},
      {"system": "sys2", "code": "code2"}
    ]},
    "system": "isys",
    "value": "value1",
    "period": {"start": "2001-05-06"},
    "assigner": {"display": "Acme Healthcare"}
   },
   {
   "use": "official",
   "type": { "coding": [
     {"system": "itsys", "code": "itcode"}
   ]},
   "system": "isys2",
   "value": "ival2"
   }
  ],
  "organization": {
     "reference": "Organization/1"
  },
  "quantity": {
     "code": "code",
     "system": "system",
     "unit": "unit"
  },
  "telecom": [
      {
        "system": "phone",
        "value": "(03) 5555 6473",
        "use": "work"
      },
      {
        "system": "email",
        "value": "ivan@mail.ru",
        "use": "home"
      }
  ]

}
$$
);
SELECT fhirpath_as_token(resource, '.identifier', 'Identifier') from as_token_test;
                 fhirpath_as_token                 
---------------------------------------------------
 {value1,isys,isys|value1,ival2,isys2,isys2|ival2}
(1 row)

SELECT fhirpath_as_token(resource, '.telecom', 'ContactPoint') from as_token_test;
                                   fhirpath_as_token                                   
---------------------------------------------------------------------------------------
 {"(03) 5555 6473",phone,"phone|(03) 5555 6473",ivan@mail.ru,email,email|ivan@mail.ru}
(1 row)

SELECT fhirpath_as_token(resource, '.telecom.value', 'string') from as_token_test;
        fhirpath_as_token        
---------------------------------
 {"(03) 5555 6473",ivan@mail.ru}
(1 row)

SELECT fhirpath_as_token(resource, '.telecom.use', 'code') from as_token_test;
 fhirpath_as_token 
-------------------
 {work,home}
(1 row)

SELECT fhirpath_as_token(resource, '.telecom.use', 'uri') from as_token_test;
 fhirpath_as_token 
-------------------
 {work,home}
(1 row)

SELECT fhirpath_as_token(resource, '.telecom.use', 'string') from as_token_test;
 fhirpath_as_token 
-------------------
 {work,home}
(1 row)

SELECT fhirpath_as_token(resource, '.identifier.type.coding', 'Coding') from as_token_test;
                            fhirpath_as_token                            
-------------------------------------------------------------------------
 {code1,sys1,sys1|code1,code2,sys2,sys2|code2,itcode,itsys,itsys|itcode}
(1 row)

SELECT fhirpath_as_token(resource, '.organization', 'Reference') from as_token_test;
 fhirpath_as_token 
-------------------
 {Organization/1}
(1 row)

SELECT fhirpath_as_token(resource, '.quantity', 'Quantity') from as_token_test;
             fhirpath_as_token              
--------------------------------------------
 {code,unit,system,system|code,system|unit}
(1 row)

SELECT fhirpath_as_token(resource, '.identifier.type', 'CodeableConcept') from as_token_test;
                            fhirpath_as_token                            
-------------------------------------------------------------------------
 {code1,sys1,sys1|code1,code2,sys2,sys2|code2,itcode,itsys,itsys|itcode}
(1 row)

SELECT fhirpath_as_token('{"a": true}', '.a', 'boolean') from as_token_test;
 fhirpath_as_token 
-------------------
 {true}
(1 row)

SELECT fhirpath_as_token('{"a": false}', '.a', 'boolean') from as_token_test;
 fhirpath_as_token 
-------------------
 {false}
(1 row)

SELECT fhirpath_as_token('{"a": "ups"}', '.a', 'boolean') from as_token_test;
 fhirpath_as_token 
-------------------
 {ups}
(1 row)

SELECT fhirpath_as_token('{"a": 1234}', '.a', 'integer') from as_token_test;
 fhirpath_as_token 
-------------------
 {1234}
(1 row)

SELECT fhirpath_as_token('{"a": 1234.1}', '.a', 'integer') from as_token_test;
 fhirpath_as_token 
-------------------
 {1234.1}
(1 row)

SELECT fhirpath_as_token('{"a": {"value": 1234}}', '.a', 'Identifier') from as_token_test;
 fhirpath_as_token 
-------------------
 {1234}
(1 row)

